### PR TITLE
Initial commit for the Harmony Hub binding for openHAB2

### PR DESCRIPTION
**Pull request by digitaldan**
Signed-off-by: Dan Cunningham <dan@digitaldan.com>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Is this currently in a state that I should start reviewing it?

**Comment by digitaldan** [comment]
Actually, hold off, I want to come back to it and review how I'm handling dynamic channels.  I'll have time this weekend to review again.

**Comment by kaikreuzer** [comment]
Ok, I'm waiting for your signal then!

**Comment by digitaldan** [comment]
@kaikreuzer I'm ready for a review!

**Comment by kaikreuzer** [comment]
Ok, thanks! I'll try to find time for it soon!

**Comment by kaikreuzer** [comment]
@digitaldan The very first thing I notice is the terrible long list of dependencies, which blows up the bundle to many MB and seem to be due to the dependencies of the harmony-java-client - we should try to reduce this set as much as possible and rather use proper OSGi bundles instead.

Since this [library is under GPL](https://github.com/tuck182/harmony-java-client/blob/master/LICENSE) we actually MUST NOT include it within openHAB, since the license is not compatible with the EPL. It seems the same lib is included in openHAB 1 already - @teichsta, did you consider this when merging the binding at the time or has this been merged unnoticed?

@tuck182, is there any chance that you would be willing to dual license the lib under GPL&EPL? This would make life a lot easier for us!

**Comment by digitaldan** [comment]
@kaikreuzer I'm all for shrinking the binding if possible.  I just removed 2 libraries (available as shared OH libraries) , the others are needed for the harmony-java-client which is the only external dependency outside of the OH libraries that I use.  Right now the binding jar is at 3.7M. 

Hopefully @tuck182 can change the licensing!

**Comment by tuck182** [comment]
I have no problem with changing the licensing. I'll get that taken care of.

**Comment by digitaldan** [comment]
Awesome, thanks!

**Comment by kaikreuzer** [comment]
Cool, thanks @tuck182!

**Comment by tuck182** [comment]
Just to update, https://github.com/tuck182/harmony-java-client has the
requested licensing changes now. Please do let me know if you need anything
more from me.

**Comment by kaikreuzer** [comment]
Many thanks, @tuck182!

@digitaldan So we now have to see what to do with the dependencies. It would be actually best to slightly refactor the harmony client lib to e.g. use GSON and other ESH "standard" libs. I am not sure about the effort though and whether this could be done as a PR against the original lib.
But usually, I would prefer a binding to be clearly <1MB, from which we are still far away...

**Comment by digitaldan** [comment]
@kaikreuzer have you read the Martian?  I'm imagining having to strip an enormous amount of heavy equipment out of the spacecraft to be able to leave the atmosphere......  Give me a while to do some research and see what can be done and how much work it's going to be.   (p.s. if you have not read it, I highly suggest you do! )

**Comment by kaikreuzer** [comment]
> @kaikreuzer have you read the Martian?

Unless somebody already submitted it as a PR or an issue, no - I have no time for reading anything else...

> Give me a while to do some research 

Yes, but don't spend too much time on it. But as long as I review the PR, you can of course look into it and see what could be optimized without too much effort :-)

**Comment by digitaldan** [comment]
@kaikreuzer sounds good, I'll take a look this week and see we I can pair back.  One question, I notice a few OH bindings use the jackson/fasterxml libraries, would this be something to consider putting in a as a shared library?  I have another binding I'm tinkering with that also has a library that depends on those as well.   I think the issue is not people needing to use the jackson stuff directly, but using libraries that have dependencies on them.

**Comment by kaikreuzer** [comment]
> would this be something to consider putting in a as a shared library?

Well, the point is that for ESH we chose GSON to be the library that mandatorily has to be available in any runtime for JSON processing. Jackson does offer more or less the same funcitonality, so adding it to the runtime at the same time blows it up without any direct benefit.
So I would not want to offer Jackson as an option for bindings to use. Just for "hard" cases like this one, we could nonetheless make it available as Karaf feature that can then be included. The risk is that as soon as Jackson is available in the target platform, people start using it...

**Comment by digitaldan** [comment]
> So I would not want to offer Jackson as an option for bindings to use. Just for "hard" cases like this one

Sounds good to me.  
This project also requires guva >= version 14 , I think OH2 is using version 10, guva is currently at v19.  Is it possible to up the guava version in OH?  this would reduce the binding size to around 1.8MB.  I wasn't sure if there is a broader reason for using v10.  I'm also unsure where that version dependency is set for the root OH2 project.

**Comment by kaikreuzer** [comment]
> Is it possible to up the guava version in OH?

Again, for ESH, we stick to version 10.1 due to size reasons, see https://github.com/eclipse/smarthome/pull/261.

> This project also requires guva >= version 14

Is this because it uses special features that are not contained in 10.1? Or because there were relevant bug fixes that are required?
One option we could think about in ESH is to move to a more recent version of Guava, while excluding packages that aren't really relevant to us from it. If you think this is worth to discuss, feel free to do so in the ESH forum (https://www.eclipse.org/forums/index.php/f/271/).

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    @Override
+    public void initialize() {
+        id = getConfig().as(HarmonyDeviceConfig.class).id;
+        name = getConfig().as(HarmonyDeviceConfig.class).name;
+
+        if (name == null && id <= 0) {
+            logger.error("A harmony device thing must be configured with a device name OR a postive device id");
+            updateStatus(ThingStatus.UNINITIALIZED, ThingStatusDetail.CONFIGURATION_ERROR);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java]: You should set it to OFFLINE / CONFIGURATION ERROR.

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        } else {
+            logName = id > 0 ? String.valueOf(id) : name;
+            logger.debug("initializing {}", logName);
+            updateStatus(ThingStatus.INITIALIZING);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java]: you must not set to INITIALIZING (this is done by the framework).

Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java]: When testing, for some reason all my devices stayed in status INITIALIZING and never went ONLINE.

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import org.eclipse.smarthome.core.thing.ThingStatus;
+
+/**
+ * the {@link HarmonyDeviceHandler} interface is for classes wishing to register
+ * to be called back when a HarmonyHub status changes
+ *
+ * @author Dan Cunningham - Initial contribution
+ *
+ */
+public interface HubStatusListener {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java]: Note that we recently introduced BaseThingHandler.bridgeStatusChanged - so I assume that you won't need this interface anymore, right? Could you please refactor your code accordingly?

File: addons/binding/pom.xml [hunk 4]
```diff
@@ -37,6 +37,7 @@
     <module>org.openhab.binding.squeezebox</module>
     <module>org.openhab.binding.tesla</module>
     <module>org.openhab.binding.vitotronic</module>
+    <module>org.openhab.binding.harmonyhub</module>
     <module>org.openhab.binding.zwave</module>
   </modules>
```
Comment (kaikreuzer) [inline addons/binding/pom.xml]: Could you please also add your binding in https://github.com/openhab/openhab2-addons/blob/master/features/openhab-addons/src/main/feature/feature.xml?

**Comment by kaikreuzer** [comment]
Hi @digitaldan,

Done my review, your code looks very good!
I'd be ok with the included libraries for now and rather want to get to a point to merge it soon. We can afterwards still try to improve the footprint of the binding by not requiring so many additional libs.

**Comment by kaikreuzer** [comment]
Hi @digitaldan,

Done my review, your code looks very good!
I'd be ok with the included libraries for now and rather want to get to a point to merge it soon. We can afterwards still try to improve the footprint of the binding by not requiring so many additional libs.

**Comment by digitaldan** [comment]
Thanks @kaikreuzer for the review! I always appreciate the time you take to review PR's.  I'm on holiday for the next few days, so expect something by the end of the week.

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java (hunk 3 above)
Comment (digitaldan) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java]: I don't need it for the child thing, but the discovery class still needs to know when the hub goes online in order to scan for new devices ( https://github.com/digitaldan/openhab2-addons/blob/harmonyhub/addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/discovery/HarmonyDeviceDiscoveryService.java )  If you know of a better way of triggering the the discovery class without this I could remove the interface and logic.

**Comment by digitaldan** [comment]
Hi Kai, except for my other comment this should be complete.  Let me know if I can squash or if you would like to see further changes. Thanks!

**Comment by kaikreuzer** [comment]
Thanks a lot! Will try to do some testing tonight and if all looks good, I'll ask you to squash then!

**Comment by kaikreuzer** [comment]
Bridge might not be available when initialize() is called, see 
```
2016-04-03 21:58:51.320 [DEBUG] [h.handler.HarmonyDeviceHandler:204  ] - Could not add button channels to device 32309796java.lang.NullPointerException: null
	at org.openhab.binding.harmonyhub.handler.HarmonyDeviceHandler.updateChannel(HarmonyDeviceHandler.java:166)
	at org.openhab.binding.harmonyhub.handler.HarmonyDeviceHandler.updateDeviceStatus(HarmonyDeviceHandler.java:146)
	at org.openhab.binding.harmonyhub.handler.HarmonyDeviceHandler.initialize(HarmonyDeviceHandler.java:111)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7$1.call(ThingManager.java:639)
```

**Comment by kaikreuzer** [comment]
Looks as if I crashed the binding by simply sending an invalid button name:
```
2016-04-03 22:02:45.770 [DEBUG] [.c.thing.internal.ThingManager:336  ] - Delegating command 'ON' for item 'harmonyhub_device_HarmonyHub_32309797_buttonPress' to handler for channel 'harmonyhub:device:HarmonyHub:32309797:buttonPress'
2016-04-03 22:02:45.772 [DEBUG] [h.handler.HarmonyDeviceHandler:89   ] - Pressing button ON on 0
Apr 03, 2016 10:02:45 PM org.jivesoftware.smack.XMPPConnection callConnectionClosedOnErrorListener
WARNUNG: Connection closed with error
net.whistlingfish.harmony.protocol.HarmonyProtocolException: Got error response [566]: Command not found for device id:32309797
	at net.whistlingfish.harmony.protocol.OAReplyProvider.parseIQ(OAReplyProvider.java:61)
	at org.jivesoftware.smack.util.PacketParserUtils.parseIQ(PacketParserUtils.java:440)
	at org.jivesoftware.smack.tcp.PacketReader.parsePackets(PacketReader.java:170)
	at org.jivesoftware.smack.tcp.PacketReader.access$000(PacketReader.java:47)
	at org.jivesoftware.smack.tcp.PacketReader$1.run(PacketReader.java:81)

2016-04-03 22:02:46.025 [ERROR] [.c.thing.internal.ThingManager:350  ] - Exception occured while calling handler: java.lang.RuntimeException: Failed communicating with Harmony Hubjava.util.concurrent.ExecutionException: java.lang.RuntimeException: Failed communicating with Harmony Hub
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.receiveCommand(ThingManager.java:339)
	at org.eclipse.smarthome.core.items.events.AbstractItemEventSubscriber.receive(AbstractItemEventSubscriber.java:46)
	at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:192)
	at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:1)
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Caused by: java.lang.RuntimeException: Failed communicating with Harmony Hub
	at net.whistlingfish.harmony.HarmonyClient.sendOAPacket(HarmonyClient.java:251)
	at net.whistlingfish.harmony.HarmonyClient.sendOAPacket(HarmonyClient.java:240)
	at net.whistlingfish.harmony.HarmonyClient.pressButton(HarmonyClient.java:342)
	at org.openhab.binding.harmonyhub.handler.HarmonyDeviceHandler.handleCommand(HarmonyDeviceHandler.java:92)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$2.call(ThingManager.java:342)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$2.call(ThingManager.java:1)
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.jivesoftware.smack.SmackException$NotConnectedException: null
	at org.jivesoftware.smack.XMPPConnection.sendPacket(XMPPConnection.java:584)
	at net.whistlingfish.harmony.HarmonyClient.sendOAPacket(HarmonyClient.java:248)
	at net.whistlingfish.harmony.HarmonyClient.sendOAPacket(HarmonyClient.java:240)
	at net.whistlingfish.harmony.HarmonyClient.pressButton(HarmonyClient.java:342)
	at org.openhab.binding.harmonyhub.handler.HarmonyDeviceHandler.handleCommand(HarmonyDeviceHandler.java:92)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$2.call(ThingManager.java:342)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$2.call(ThingManager.java:1)
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
```
There should only be a single logging line (warning?) in such a case and the XMPP connection should not be disconnected through it.

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java [hunk 5]
```diff
@@ -102,12 +102,13 @@ public void handleCommand(ChannelUID channelUID, Command command) {
     public void initialize() {
         id = getConfig().as(HarmonyDeviceConfig.class).id;
         name = getConfig().as(HarmonyDeviceConfig.class).name;
-        if (name == null && id <= 0) {
+        if (!checkConfig()) {
             logger.error("A harmony device thing must be configured with a device name OR a postive device id");
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java]: don't log any error here, but add the message to the updateStatus call as a third parameter please!

**Comment by kaikreuzer** [comment]
Ok, as you can see above, I did some tests - it feels as if the binding should be further thoroughly tested, but I guess this will happen anyhow after merging it :-)
If you could address the issues mentioned above, it would be nice, though.
Before squashing&merging, I have a very last wish: Could you please also add [an about.html file](https://eclipse.org/legal/epl/longabout.html), which lists the included libraries and their licenses? See https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.ipp/about.html#L28 as an example.

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java (hunk 3 above)
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java]: I am not sure whether it is a good idea to trigger a scan on every status change to ONLINE. This should really only be done if it is a "cheap" operation. But at first initialization, it should indeed be done and I do not know any better way to trigger the scan than using a listener construct...

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java (hunk 5 above)
Comment (digitaldan) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HarmonyDeviceHandler.java]: Will do

File: addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java (hunk 3 above)
Comment (digitaldan) [inline addons/binding/org.openhab.binding.harmonyhub/src/main/java/org/openhab/binding/harmonyhub/handler/HubStatusListener.java]: Fortunately a scan is very cheap, all it does is ask the bridge/hub for a list of devices.

**Comment by digitaldan** [comment]
Thanks @kaikreuzer , I will look at trying to fix the disconnect issue in the harmony library and add the library about file.   As for testing, I have had it running with 4 hubs for a couple of months and had a couple of other users that have been trying it out and giving feedback, so I'm fairly confident it should be ok.

**Comment by kaikreuzer** [comment]
What do you need 4 hubs for...? Ok, glad to hear that it works well already - it just seems me who is always able to get exceptions everywhere when playing with things :-D

**Comment by digitaldan** [comment]
I guess I just really like Harmony Hubs ;-)

**Comment by kaikreuzer** [comment]
Hey, @digitaldan, any update here? I'd love to merge this!

**Comment by digitaldan** [comment]
Hi Kai, I will try and get this done later tonight or tomorrow!

On Fri, Apr 15, 2016 at 8:09 AM Kai Kreuzer <notifications@github.com>
wrote:

> Hey, @digitaldan <https://github.com/digitaldan>, any update here? I'd
> love to merge this!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/openhab/openhab2-addons/pull/639#issuecomment-210498439>
>
-- 
Sent from my mobile.

**Comment by kaikreuzer** [comment]
@digitaldan Are you sure...?

**Comment by digitaldan** [comment]
Hi @kaikreuzer,  I updated the lib to handle the recoverable 566 error and made changes according to your comments.  The underlying xmpp library is still very chatty about errors, but fortunately we reconnects when this happens.

**Comment by kaikreuzer** [comment]
That's awesome that you found the time for that despite all your activity on the native apps!
Many thanks, let's merge!

**Pull request merged**